### PR TITLE
Reproduce failed breakout detection in multistage queries

### PR DESCRIPTION
**Pull request by iethree**
reproduces https://github.com/metabase/metabase/issues/22563

### Problem

The reproduction is complex in the issue, but the essence of this bug is that when you have custom columns after aggregations in a query, the query becomes multistage and the calls to `columnDimensions()` in `/metabase-lib/queries/utils/dataset.js` don't treat the query as having aggregations because it only checks the last stage.  

*(you can check this by following the repro, but put the custom column before the aggregation and see that everything works fine)*

This creates two problems: 
- First, we show `additionalOptions` in the column settings, which are nonsensical for aggregated queries.
- Second, we get `columnDimensions` that don't follow the correct logic path because we ignore the aggregations, which causes `syncTableColumnsToQuery()` in `/metabase-lib/utils/dataset.js` to get inconsistent results from `columnDimensions` and it thinks it needs to add a fieldsClause to the query, which is what messes up the columns to create this bug.

## Solution

Wait for MLv2 to save us with its better handling of query stages. Commiting the reproduction for now.

~~The fix I found was to check not just the top level query for breakouts, but all stages of the query both in producing `fieldOptions()` and `columnDimensions()`)~~

Before | After
--- | ---
beforebreakoutsfix | afteraggfix

~~This is accomplished by adding a new `hasAnyBreakouts()` function to the StructuredQuery class that checks all query stages for breakouts, and using that in `fieldOptions()` and `columnDimensions()`. ~~

~~This certainly solves this issue, but what is less clear is whether these changes to `columnDimensions` could have side effects elsewhere. 🤔  It is only called in `dataset.js`, but that affects nearly every query we send to the backend.~~

### How to verify

Follow Flamber's repro steps, or run the e2e test.

### CODE REVIEW

File: frontend/src/metabase-lib/queries/StructuredQuery.ts [hunk 1]
```diff
@@ -1192,7 +1196,7 @@ class StructuredQueryInner extends AtomicQuery {
   fieldsOptions(
     dimensionFilter: DimensionFilterFn = dimension => true,
   ): DimensionOptions {
-    if (this.isBareRows() && !this.hasBreakouts()) {
+    if (this.isBareRows() && !this.hasAnyBreakouts()) {
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/queries/StructuredQuery.ts:1199]: This line did not change semantically but behavior did.
It's not possible to guess/tell the difference between `hasBreakouts()` and `hasAnyBreakouts()` without looking at their code. I suggest to make the name more specific:


```suggestion
    if (this.isBareRows() && !this.anySourceQueryHasBreakouts()) {
```

File: frontend/src/metabase-lib/queries/StructuredQuery.ts [hunk 2]
```diff
@@ -538,6 +538,10 @@ class StructuredQueryInner extends AtomicQuery {
     return this.breakouts().length > 0;
   }
 
+  hasAnyBreakouts() {
+    return this.queries().some(query => query.hasBreakouts());
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/queries/StructuredQuery.ts:542]: Should we also check `this.query()` here?

I mean

```suggestion
    return [this.query(), ...this.queries()].some(query => query.hasBreakouts());
```

or

```suggestion
    return this.hasBreakouts() || this.queries().some(query => query.hasBreakouts());
```

----

If that's the case, then I'd call this function `anyQueryHasBreakouts`.

**Review by ranquild — CHANGES_REQUESTED** [review]
@iethree We're currently migrating this part of the app to MLv2 here https://github.com/metabase/metabase/pull/32698. I think it would make sense for the MLv2 PR to land and then fix the issue on the clojure side if it exists. Wdyt?

**Comment by iethree** [comment]
> @iethree We're currently migrating this part of the app to MLv2 here #32698. I think it would make sense for the MLv2 PR to land and then fix the issue on the clojure side if it exists. Wdyt?

how about i change this PR to just have the skipped repro then?

**Review by kamilmielnik — APPROVED** [review]
:+1:

**Review by ranquild — APPROVED** [review]

**Pull request merged**